### PR TITLE
server/perkeepd/ui: Implement blob audio

### PR DESCRIPTION
**Pull request by omarchehab98**
Adds audio support to the web UI

### CODE REVIEW

File: server/perkeepd/ui/blob_item_audio_content.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+		blobref: React.PropTypes.string.isRequired,
+		filename: React.PropTypes.string.isRequired,
+		href: React.PropTypes.string.isRequired,
+		size: React.PropTypes.instanceOf(goog.math.Size).isRequired,
+	},
+
+	getInitialState: function() {
+		return {
+			mouseover: false,
+			playing: false,
```
Comment (omarchehab98) [inline server/perkeepd/ui/blob_item_audio_content.js]: @mpl in response to your comments on Gerrit:

>I think the point was not about the media being loaded, but about the react element itself. imagine there are 50 videos in your current page, then it probably takes a little bit of time for all the buttons to load on the page.
>[Source](https://perkeep-review.googlesource.com/c/perkeep/+/19326/1/server/perkeepd/ui/blob_item_audio_content.js#37)

Since we're using React, buttons are created and added to the DOM on the client-side. It's a safe assumption that your buttons will be rendered soon after your React component is instantiated.

Correction, I said `loaded` is not used in video preview, it is, here is how `loaded` is used in `blob_item_video_content`:
* Adds the class `cam-blobitem-video-loaded` if `loaded`, however, class is not implemented
* Renders `getPoster_` if not `loaded`
* Renders `getVideo_` if `loaded`
* Sets `loaded` to `true` when clicking play or pause button

Basically, loaded is whether or not the video was ever played by the user. We use that information to decide whether to display the poster or the video. Audio has no visual change when it is being played, so we don't need `loaded` here.

File: server/perkeepd/ui/blob_item_audio_content.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+			mouseover: false,
+		});
+	},
+});
+
+cam.BlobItemAudioContent.isAudio = function(rm) {
+	// From https://developer.mozilla.org/en-US/docs/Web/HTML/Supported_media_formats
+	var extensions = [
+		'webm',
+		'ogg',
```
Comment (omarchehab98) [inline server/perkeepd/ui/blob_item_audio_content.js]: >>Good catch, I didn't know `ogg` is also a video format. Let's ensure video takes precedence
>
>Because you added the audio handler after the video one in BLOB_ITEM_HANDLERS_, I think it's already the case.
>
>>and use metadata to determine whether the blob is audio or video.
>
>yes, that would be nice.
>But on first approximation if you want, we can assume most oggs are in fact audio (which I believe is practically the case), and the video handler can always let the audio handler deal with oggs for now.

TODO: Give audio priority over video for `ogg`

File: server/perkeepd/ui/blob_item_audio_content.js (hunk 1 above)
Comment (mpl) [reply server/perkeepd/ui/blob_item_audio_content.js]: understood.

File: server/perkeepd/ui/blob_item_audio_content.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+		);
+	},
+
+	setAudioRef_: function(audio) {
+		this.audioRef_ = audio;
+	},
+
+	handlePlayPauseClick_: function(e) {
+		if (this.state.playing) {
+			this.audioRef_.pause();
```
Comment (mpl) [inline server/perkeepd/ui/blob_item_audio_content.js]: why is it better to keep an audioRef_ around, rather than use ReactDOM.findDOMNode(this).querySelector('audio') to find it, like it was done for video?

File: server/perkeepd/ui/blob_item_audio.css [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	text-align: center;
+}
+
+.cam-blobitem-audio-play,
+.cam-blobitem-audio-pause {
+	color: rgba(125, 125, 125, 0.7);
+	position: absolute;
+	left: 50%;
+	top: 50%;
+	transform: translate(-50%, -50%);
```
Comment (mpl) [inline server/perkeepd/ui/blob_item_audio.css]: what does that do?

File: server/perkeepd/ui/index.js [hunk 5]
```diff
@@ -80,6 +81,7 @@ cam.IndexPage = React.createClass({
 		cam.BlobItemInstapaperContent.getHandler,
 		cam.BlobItemImageContent.getHandler,
 		cam.BlobItemVideoContent.getHandler,
+		cam.BlobItemAudioContent.getHandler,
```
Comment (mpl) [inline server/perkeepd/ui/index.js]: Just move the audio handler above the video handler, and I think that should be enough to give the audio handler priority for oggs.

**Comment by omarchehab98** [comment]
findDOMNode is deprecated, using it discouraged

File: server/perkeepd/ui/audio_detail.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+goog.provide('cam.AudioDetail');
+
+goog.require('cam.BlobItemAudioContent');
+
+// Renders the guts of the detail view for images.
```
Comment (mpl) [inline server/perkeepd/ui/audio_detail.js]: comment needs fixing

File: server/perkeepd/ui/audio_detail.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+cam.AudioDetail = React.createClass({
+	displayName: 'AudioDetail',
+
+	propTypes: {
+		backwardPiggy: React.PropTypes.bool.isRequired,
```
Comment (mpl) [inline server/perkeepd/ui/audio_detail.js]: we don't need the piggy for this aspect, do we?

File: server/perkeepd/ui/audio_detail.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+		height: React.PropTypes.number.isRequired,
```
Comment (mpl) [inline server/perkeepd/ui/audio_detail.js]: height does not seem to be used.

File: server/perkeepd/ui/audio_detail.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+		permanodeMeta: React.PropTypes.object,
+		resolvedMeta: React.PropTypes.object.isRequired,
+		width: React.PropTypes.number.isRequired,
```
Comment (mpl) [inline server/perkeepd/ui/audio_detail.js]: same as for weight

File: server/perkeepd/ui/audio_detail.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+				},
+				React.DOM.audio({
+					controls: true,
+					key: this.props.resolvedMeta.blobRef,
+					src: goog.string.subs('%s%s/%s', goog.global.CAMLISTORE_CONFIG.downloadHelper, this.props.resolvedMeta.blobRef, this.props.resolvedMeta.file.fileName),
+				}),
+			),
+		);
+	},
+});
```
Comment (mpl) [inline server/perkeepd/ui/audio_detail.js]: this looks nice, thanks.

File: server/perkeepd/ui/blob_item_audio.css (hunk 4 above)
Comment (omarchehab98) [reply server/perkeepd/ui/blob_item_audio.css]: `left: 50%` and `top: 50%` move the element's top left corner to the center `transform: translate(-50%, -50%)` moves the element 50% of it's width and height towards to the negative `x` and `y` direction.

In english, it centers the element using the element's center as an anchor

Comment (mpl) [reply server/perkeepd/ui/blob_item_audio.css]: ACK, thx.

**Comment by omarchehab98** [comment]
@mpl sounds good 🚀 

Regarding force push, if you click on the link "force-pushed" it will show a diff of what got amended
Alternatively, if you look at the files changed tab you will be able to see all the changes with the new amends
